pkg/bugtool/maps.go

```diff
@@ -0,0 +1,355 @@
+		info, err := m.Info()
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve map info: %w", err)
+		}
+
+		memlock, err := parseMemlockFromFDInfo(m.FD())
+		if err != nil {
+			return nil, fmt.Errorf("failed parsing fdinfo to retrieve memlock: %w", err)
+		}
+
+		mapInfos = append(mapInfos, ExtendedMapInfo{
+			MapInfo: *info,
+			Memlock: memlock,
+		})
+	}
+}
+
+// FindPinnedMaps returns all info of maps pinned under the path
+// specified as argument.
+func FindPinnedMaps(path string) ([]ExtendedMapInfo, error) {
+	var infos []ExtendedMapInfo
+	err := filepath.WalkDir(path, func(path string, d fs.DirEntry, _ error) error {
+		if d.IsDir() {
+			return nil // skip directories
+		}
+		m, err := ebpf.LoadPinnedMap(path, &ebpf.LoadPinOptions{
+			ReadOnly: true,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to load pinned map %q: %w", path, err)
+		}
+		defer m.Close()
+
+		// check if it's really a map because ebpf.LoadPinnedMap does not return
+		// an error but garbage info on doing this on a prog
+		if ok, err := isMap(m.FD()); err != nil || !ok {
+			if err != nil {
+				return err
+			}
+			return nil // skip non map
+		}
+
+		info, err := m.Info()
+		if err != nil {
+			return fmt.Errorf("failed to retrieve map info: %w", err)
+		}
+
+		memlock, err := parseMemlockFromFDInfo(m.FD())
+		if err != nil {
+			return fmt.Errorf("failed to parse memlock from fd (%d) info: %w", m.FD(), err)
+		}
+
+		infos = append(infos, ExtendedMapInfo{
+			MapInfo: *info,
+			Memlock: memlock,
+		})
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return infos, nil
+}
+
+// mapIDsFromProgs retrieves all map IDs used inside a prog.
+func mapIDsFromProgs(prog *ebpf.Program) ([]int, error) {
+	if prog == nil {
+		return nil, fmt.Errorf("prog is nil")
+	}
+	progInfo, err := prog.Info()
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve prog info: %w", err)
+	}
+	// check if field is available
+	ids, available := progInfo.MapIDs()
+	if !available {
+		return nil, fmt.Errorf("can't link prog to map IDs, field available from 4.15")
+	}
+	mapSet := map[int]bool{}
+	for _, id := range ids {
+		mapSet[int(id)] = true
+	}
+	return maps.Keys(mapSet), nil
+}
+
+// mapIDsFromPinnedProgs scan the given path and returns the map IDs used by the
+// prog pinned under the path. It also retrieves the map IDs used by the prog
+// referenced by program array maps (tail calls). This should only work from
+// kernel 4.15.
+func mapIDsFromPinnedProgs(path string) ([]int, error) {
+	mapSet := map[int]bool{}
+	progArrays := []*ebpf.Map{}
+	err := filepath.WalkDir(path, func(path string, d fs.DirEntry, _ error) error {
+		if d.IsDir() {
+			return nil // skip directories
+		}
+		prog, err := ebpf.LoadPinnedProgram(path, &ebpf.LoadPinOptions{
+			ReadOnly: true,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to load pinned object %q: %w", path, err)
+		}
+		defer prog.Close()
+
+		if ok, err := isProg(prog.FD()); err != nil || !ok {
```

Comment:
fyi I was using filename check for 'prog' or 'prog_override' to identify programs, your's is more resilient

Reply:
ok thanks didn't know! I think I'll integrate that into c/ebpf eventually https://github.com/cilium/ebpf/issues/1566.